File: examples/src/main/java/com/freya02/bot/wiki/localization/commands/SlashBan.java
package com.freya02.bot.wiki.localization.commands;

import com.freya02.botcommands.api.application.ApplicationCommand;
import com.freya02.botcommands.api.application.annotations.AppOption;
import com.freya02.botcommands.api.application.slash.GuildSlashEvent;
import com.freya02.botcommands.api.application.slash.annotations.JDASlashCommand;
import com.freya02.botcommands.api.localization.DefaultLocalizationTemplate;
import com.freya02.botcommands.api.localization.annotations.LocalizationBundle;
import com.freya02.botcommands.api.localization.providers.DefaultLocalizationMapProvider;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.interactions.commands.localization.LocalizationFunction;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.concurrent.TimeUnit;

import static com.freya02.botcommands.api.localization.Localization.Entry.entry;

//TODO waiting on JDA localization changes (#2261)
/**
 * A few resources:
 *
 * <ul>
 *     <li>{@link DefaultLocalizationMapProvider DefaultLocalizationMapProvider - Where to put your localization file}</li>
 *     <li>{@link DefaultLocalizationTemplate DefaultLocalizationTemplate - How localization templates are formatted}</li>
 *     <li>{@link LocalizationFunction LocalizationFunction - How localization keys are made (from JDA)}</li>
 * </ul>
 */
@LocalizationBundle("LocalizationWikiCommands") //The file is in /resources/bc_localization/LocalizationWikiCommands.json
public class SlashBan extends ApplicationCommand {
	//Description is set in localization
	@JDASlashCommand(name = "ban")
	public void onSlashBan(GuildSlashEvent event,
	                       @AppOption(name = "user") User targetUser, //Description is set in localization
	                       @AppOption int delHours, //Description is set in localization
	                       @AppOption @Nullable String reason) { //Description is set in localization

		event.deferReply(true).queue();

		//Following localized strings will appear in the user's language, as it is provided for interactions (slash commands, context, buttons...)

		final Member caller = event.getMember();
		final Member botMember = event.getGuild().getSelfMember();

		//Check bot permissions
		if (!botMember.hasPermission(Permission.BAN_MEMBERS)) {
			final String localize = event.localize("ban.bot.permission_error");
			event.getHook().sendMessage(localize).queue();
			return;
		}

		//Check permissions & hierarchy
		// This requires the JDA chunking filter and member cache policy to be enabled, as this uses member cache.
		final Member targetMember = event.getGuild().getMember(targetUser);
		if (targetMember != null) { //Can be null if caching isn't set up
			if (!caller.canInteract(targetMember)) { //Check if the caller can interact with the banned member
				event.getHook()
						.sendMessage(event.localize("ban.caller.interact_error", entry("mention", targetMember.getAsMention())))
						.queue();
				return;
			} else if (!caller.hasPermission(Permission.BAN_MEMBERS)) { //Check if the caller can ban
				event.getHook()
						.sendMessage(event.localize("ban.caller.permission_error"))
						.queue();
				return;
			} else if (!botMember.canInteract(targetMember)) { //Check if the bot can interact with the banned member
				event.getHook()
						.sendMessage(event.localize("ban.bot.interact_error", entry("mention", targetMember.getAsMention())))
						.queue();
				return;
			}
		}

		final String notNullReason = Objects.requireNonNullElse(reason, "No reason");
		final String logReason = "Banned by " + caller.getAsMention() + " : " + notNullReason;

		//Ban user
		event.getGuild()
				.ban(targetUser, delHours, TimeUnit.HOURS)
				.reason(logReason)
				.queue(x -> {
					event.getHook().sendMessage(event.localize("ban.success",
							entry("banned_user", targetUser.getAsMention()),
							entry("del_hours", delHours),
							entry("reason", notNullReason)
					)).queue();
				}, t -> {
					event.getHook().sendMessage(event.localize("ban.unknown_error")).queue();
				});
	}
}

File: BotTemplate/BotTemplate-Kotlin/src/main/kotlin/io/github/freya022/bot/commands/slash/SlashBan.kt
package io.github.freya022.bot.commands.slash

import com.freya02.botcommands.api.commands.annotations.BotPermissions
import com.freya02.botcommands.api.commands.annotations.Command
import com.freya02.botcommands.api.commands.annotations.UserPermissions
import com.freya02.botcommands.api.commands.application.ApplicationCommand
import com.freya02.botcommands.api.commands.application.GlobalApplicationCommandManager
import com.freya02.botcommands.api.commands.application.annotations.AppDeclaration
import com.freya02.botcommands.api.commands.application.slash.GuildSlashEvent
import com.freya02.botcommands.api.commands.application.slash.annotations.JDASlashCommand
import com.freya02.botcommands.api.commands.application.slash.annotations.SlashOption
import com.freya02.botcommands.api.components.Components
import com.freya02.botcommands.api.components.awaitAny
import com.freya02.botcommands.api.components.event.ButtonEvent
import com.freya02.botcommands.api.core.entities.InputUser
import com.freya02.botcommands.api.core.service.annotations.BService
import com.freya02.botcommands.api.core.service.annotations.ConditionalService
import com.freya02.botcommands.api.core.utils.delay
import com.freya02.botcommands.api.localization.annotations.LocalizationBundle
import com.freya02.botcommands.api.localization.context.AppLocalizationContext
import com.freya02.botcommands.api.localization.context.editLocalized
import com.freya02.botcommands.api.localization.context.replaceLocalized
import com.freya02.botcommands.api.localization.context.replyLocalizedEphemeral
import io.github.freya022.bot.commands.FrontendChooser
import io.github.freya022.bot.commands.SimpleFrontend
import io.github.freya022.bot.resolvers.localize
import kotlinx.coroutines.TimeoutCancellationException
import mu.KotlinLogging
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle
import java.util.concurrent.TimeUnit
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds

private val logger = KotlinLogging.logger { }

// This data class is practically pointless;
// this is just to demonstrate how you can group parameters together,
// so you can benefit from functions/backed properties limited to your parameters,
// without polluting classes with extensions
data class DeleteTimeframe(val time: Long, val unit: TimeUnit) {
    override fun toString(): String = "$time ${unit.name.lowercase()}"
}

@BService
class SlashBan(private val componentsService: Components) {
    suspend fun onSlashBan(
        event: GuildSlashEvent,
        @LocalizationBundle("Commands", prefix = "ban") localizationContext: AppLocalizationContext,
        target: InputUser,
        timeframe: DeleteTimeframe,
        reason: String = localizationContext.localize("outputs.defaultReason")
    ) {
        target.member?.let { targetMember ->
            if (!event.member.canInteract(targetMember)) {
                return event.replyLocalizedEphemeral(localizationContext, "errors.user.interactError", "userMention" to target.asMention).queue()
            } else if (!event.guild.selfMember.canInteract(targetMember)) {
                return event.replyLocalizedEphemeral(localizationContext, "errors.bot.interactError", "userMention" to target.asMention).queue()
            }
        }

        val cancelButton = componentsService.ephemeralButton(ButtonStyle.PRIMARY, localizationContext.localize("buttons.cancel")) {
            oneUse = true
            // Restrict button to caller, not necessary since this is an ephemeral reply tho
            constraints += event.user
        }
        val confirmButton = componentsService.ephemeralButton(ButtonStyle.DANGER, localizationContext.localize("buttons.confirm")) {
            oneUse = true
            // Restrict button to caller, not necessary since this is an ephemeral reply tho
            constraints += event.user
        }

        val componentGroup = componentsService.newEphemeralGroup(cancelButton, confirmButton) {
            timeout(1.minutes)
        }

        event.replyLocalizedEphemeral(localizationContext, "outputs.confirmationMessage", "userMention" to target.asMention)
            .addActionRow(cancelButton, confirmButton)
            .queue()

        val componentEvent: ButtonEvent = try {
            componentGroup.awaitAny()
        } catch (e: TimeoutCancellationException) {
            return event.hook.editLocalized(localizationContext, "outputs.timeout")
                .delay(5.seconds)
                .flatMap { event.hook.deleteOriginal() }
                .queue()
        }

        when (componentEvent.componentId) {
            cancelButton.id -> {
                logger.debug { "Ban cancelled for ${target.id}" }
                componentEvent.replaceLocalized(localizationContext, "outputs.cancelled").queue()

                //Cancel logic
            }
            confirmButton.id -> {
                logger.debug { "Ban confirmed for ${target.id}, $timeframe of messages were deleted, reason: '$reason'" }

                componentEvent.replaceLocalized(
                    localizationContext,
                    "outputs.success",
                    "userMention" to target.asMention,
                    "time" to timeframe.time,
                    "unit" to timeframe.unit.localize(timeframe.time, localizationContext),
                    "reason" to reason
                ).queue()

                //Ban logic
            }
            else -> throw IllegalArgumentException("Unknown button ID: ${componentEvent.componentId}")
        }
    }
}

@Command
@ConditionalService(FrontendChooser::class)
class SlashBanDetailedFront {
    @AppDeclaration
    fun onDeclare(manager: GlobalApplicationCommandManager) {
        manager.slashCommand("ban", function = SlashBan::onSlashBan) {
            description = "Ban any user from this guild"

            botPermissions += Permission.BAN_MEMBERS
            userPermissions += Permission.BAN_MEMBERS

            customOption("localizationContext")

            option("target") {
                description = "The user to ban"
            }

            aggregate("timeframe", ::DeleteTimeframe) {
                option("time") {
                    description = "The timeframe of messages to delete with the specified unit"
                }

                option("unit") {
                    description = "The unit of the delete timeframe"

                    usePredefinedChoices = true
                }
            }

            option("reason") {
                description = "The reason for the ban"
            }
        }
    }
}

@Command
@SimpleFrontend
@ConditionalService(FrontendChooser::class)
class SlashBanSimplifiedFront(private val banImpl: SlashBan) : ApplicationCommand() {
    @UserPermissions(Permission.BAN_MEMBERS)
    @BotPermissions(Permission.BAN_MEMBERS)
    @JDASlashCommand(name = "ban", description = "Ban any user from this guild")
    suspend fun onSlashBan(
        event: GuildSlashEvent,
        @LocalizationBundle("Commands", prefix = "ban") localizationContext: AppLocalizationContext,
        @SlashOption(description = "The user to ban") target: InputUser,
        @SlashOption(description = "The timeframe of messages to delete with the specified unit") time: Long,
        @SlashOption(description = "The unit of the delete timeframe", usePredefinedChoices = true) unit: TimeUnit,
        @SlashOption(description = "The reason for the ban") reason: String = localizationContext.localize("outputs.defaultReason")
    ) = banImpl.onSlashBan(event, localizationContext, target, DeleteTimeframe(time, unit), reason)
}